azurelinuxagent/ga/update.py

```diff
@@ -491,45 +496,93 @@ def _check_threads_running(self, all_thread_handlers):
+                    break
 
-            protocol.client.update_goal_state(silent=self._update_goal_state_error_count >= max_errors_to_log, save_to_history=True)
+                if self._check_certificates(goal_state):
+                    if attempt > 0:
+                        event.info(WALAEventOperation.FetchGoalState, "The extensions goal state is now in sync with the tenant cert.")
+                    break
+
+                if attempt == 0:
+                    pass  # retry without waiting
+                elif attempt == 1:
+                    time.sleep(conf.get_goal_state_period())
+                else:
+                    event.warn(WALAEventOperation.FetchGoalState, "The extensions are still out of sync with the tenant cert. Will continue execution, but some extensions may fail.")
+                    break
 
             self._goal_state = protocol.get_goal_state()
 
             if self._update_goal_state_error_count > 0:
-                message = u"Fetching the goal state recovered from previous errors. Fetched {0} (certificates: {1})".format(
+                event.info(
+                    WALAEventOperation.FetchGoalState,
+                    "Fetching the goal state recovered from previous errors. Fetched {0} (certificates: {1})",
                     self._goal_state.extensions_goal_state.id, self._goal_state.certs.summary)
-                add_event(AGENT_NAME, op=WALAEventOperation.FetchGoalState, version=CURRENT_VERSION, is_success=True, message=message, log_event=False)
-                logger.info(message)
                 self._update_goal_state_error_count = 0
 
             try:
                 self._supports_fast_track = conf.get_enable_fast_track() and protocol.client.get_host_plugin().check_vm_settings_support()
             except VmSettingsNotSupported:
                 self._supports_fast_track = False
 
+            return True
+
         except Exception as e:
             self._update_goal_state_error_count += 1
             self._heartbeat_update_goal_state_error_count += 1
             if self._update_goal_state_error_count <= max_errors_to_log:
-                message = u"Error fetching the goal state: {0}".format(textutil.format_exception(e))
-                logger.error(message)
-                add_event(op=WALAEventOperation.FetchGoalState, is_success=False, message=message, log_event=False)
-                self._update_goal_state_last_error_report = datetime.now()
+                # Report up to 'max_errors_to_log' immediately
+                self._update_goal_state_next_error_report = datetime.now()
+                event.error(WALAEventOperation.FetchGoalState, "Error fetching the goal state: {0}", textutil.format_exception(e))
             else:
-                if self._update_goal_state_last_error_report + timedelta(hours=6) > datetime.now():
-                    self._update_goal_state_last_error_report = datetime.now()
```

Comment:
while working on these fixes, I noticed that these messages were being logged every 6 secs, rather than every 6 hours as was intended.